lib/demuxer.ex

```diff
@@ -0,0 +1,175 @@
+  end
+
+  @impl true
+  def handle_demand(Pad.ref(:output, _id), _size, :buffers, context, state)
+      when state.phase == :all_outputs_linked do
+    process_actions(state.actions_buffer, context, %State{state | actions_buffer: []})
+  end
+
+  @impl true
+  def handle_demand(Pad.ref(:output, _id), _size, :buffers, _context, state) do
+    {:ok, state}
+  end
+
+  defp process_actions(actions, context, state) do
+    demands = get_demands_from_context(context, state)
+
+    {sent_actions, buffered_actions} = classify_actions(actions, demands, state)
+
+    state = %State{state | actions_buffer: buffered_actions}
+
+    actions = sent_actions ++ demand_if_not_blocked(state)
+
+    {actions, state}
+  end
+
+  defp get_demands_from_context(context, state) do
+    Enum.reduce(state.tracks, %{}, fn track, acc ->
+      Map.put(acc, track, context.pads[Pad.ref(:output, track)].demand)
+    end)
+  end
+
+  defp process_packets(packets_list, state) do
+    Enum.reduce(packets_list, {[], state}, fn packet, {actions, state} ->
+      {new_actions, state} = process_packet(packet, state)
+      {actions ++ new_actions, state}
+    end)
+  end
+
+  defp process_packet(packet, state) do
+    cond do
+      packet.bos? -> process_bos_packet(packet, state)
+      packet.eos? and packet.payload == <<>> -> {[], state}
+      true -> process_data_packet(packet, state)
+    end
+  end
+
+  defp process_bos_packet(
+         %{payload: <<"OpusHead", _rest::binary>>} = packet,
+         state
+       ) do
+    new_track_action = {:notify_parent, {:new_track, {packet.track_id, :opus}}}
+
+    {[new_track_action], %State{state | phase: :awaiting_linking}}
+  end
+
+  defp process_bos_packet(_other, _state) do
+    raise "Invalid bos packet, probably unsupported codec."
+  end
+
+  defp process_data_packet(%{payload: <<"OpusTags", _rest::binary>>}, state) do
+    {[], state}
+  end
+
+  defp process_data_packet(packet, state) do
+    buffer_action =
+      {:buffer,
+       {Pad.ref(:output, packet.track_id),
+        %Buffer{
+          payload: packet.payload
+        }}}
+
+    {[buffer_action], state}
+  end
+
+  defp demand_if_not_blocked(state) do
+    if blocked?(state), do: [], else: [demand: :input]
+  end
+
+  defp blocked?(state) do
+    not Enum.empty?(state.actions_buffer) or state.phase != :all_outputs_linked
+  end
+
+  defp classify_buffer_action(
+         {:buffer, {Pad.ref(:output, id), _buffer}} = buffer_action,
+         {sent_actions, cached_actions, demands}
+       ) do
+    if demands[id] > 0 and Enum.empty?(cached_actions) do
+      demands = %{demands | id => demands[id] - 1}
+      {sent_actions ++ [buffer_action], cached_actions, demands}
+    else
+      {sent_actions, cached_actions ++ [buffer_action], demands}
+    end
+  end
+
+  defp classify_actions(actions, demands, state) do
+    if blocked?(state) do
+      Enum.split_with(actions, fn
+        {:buffer, _} -> false
+        {_, _} -> true
```

Comment:
I believe this should be  
```suggestion
        {:notify_parent, {:new_track, _}} -> true
        _other -> false
```